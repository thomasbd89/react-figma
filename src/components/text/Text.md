# Text

Wrapper for Figma [Text](https://www.figma.com/plugin-docs/api/TextNode/).
`Text` component supports auto loading fonts defined through `fontFamily`, `fontWeight`, `fontStyle` prop. 

#### Props

| Prop       | Type     | Default | Note                                              |
| ---------- | -------- | ------- | ------------------------------------------------- |
| `name`     | `String` |         | The name to be displayed in the Figma Layers List |
| `children` | `String` |         | Text content                                      |
| `style`    | [`Style`](/docs/styling.md#type-styles)   |         |                  |

Also, most of [TextNode](https://www.figma.com/plugin-docs/api/TextNode/) fields supported as props.

#### Example

```javascript
<Text
  style={{
    fontSize: 24,
    fontFamily: 'Helvetica',
    fontWeight: 'bold',
    color: '#01ffae',
  }}
>
  Hello World!
</Text>
```
