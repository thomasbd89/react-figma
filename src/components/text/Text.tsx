import * as React from 'react';
import { DefaultShapeProps, StyleOf, TextNodeProps } from '../../types';
import {
    LayoutStyleProperties,
    transformLayoutStyleProperties
} from '../../styleTransformers/transformLayoutStyleProperties';
import {
    TextStyleProperties,
    transformTextStyleProperties
} from '../../styleTransformers/transformTextStyleProperties';
import { useYogaLayout } from '../../hooks/useYogaLayout';
import { transformBlendProperties, BlendStyleProperties } from '../../styleTransformers/transformBlendProperties';
import { YogaStyleProperties } from '../../yoga/YogaStyleProperties';
import { StyleSheet } from '../..';

export interface TextProps extends TextNodeProps, DefaultShapeProps {
    style?: StyleOf<YogaStyleProperties & LayoutStyleProperties & TextStyleProperties & BlendStyleProperties>;
    children?: string;
}

export const Text: React.FC<TextProps> = props => {
    const yogaRef = React.useRef();

    const style = StyleSheet.flatten(props.style);

    const textProps = {
        ...transformLayoutStyleProperties(style),
        ...transformTextStyleProperties(style),
        ...transformBlendProperties(style),
        ...props
    };
    const yogaProps = useYogaLayout({ yogaRef, ...textProps });
    // @ts-ignore
    return <text {...textProps} {...yogaProps} innerRef={yogaRef} />;
};
