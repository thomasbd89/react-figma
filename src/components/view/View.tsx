import * as React from 'react';
import { Frame, Rectangle, StyleSheet } from '../..';
import { RectangleProps } from '../rectangle/Rectangle';
import { FrameNodeProps } from '../frame/Frame';

export type ViewProps = FrameNodeProps | RectangleProps;

export const View: React.FC<ViewProps> = (props) => {
    if (props.children) {
        return (
            <Frame
                {...props}
                style={[{ backgroundColor: 'transparent' }, props.style && StyleSheet.flatten(props.style)]}
            />
        );
    } else {
        return <Rectangle {...(props as RectangleProps)} />;
    }
};
