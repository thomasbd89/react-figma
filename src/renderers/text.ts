import { colorToRGB } from '../helpers/color';
import { baseNodeMixin } from '../mixins/baseNodeMixin';

export const text = async props => {
    const textNode = figma.createText();

    baseNodeMixin(textNode)(props);

    let fontName = textNode.fontName;
    if (typeof fontName !== 'object') {
        fontName = {
            family: 'Roboto',
            style: 'Regular'
        };
    }
    await figma.loadFontAsync(fontName);

    textNode.characters = props.children;

    const { color } = props.style;
    if (color) {
        textNode.fills = [{ type: 'SOLID', color: colorToRGB(color) }];
    }

    figma.currentPage.appendChild(textNode);
};
