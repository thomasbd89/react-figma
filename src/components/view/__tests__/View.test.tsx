import * as React from 'react';
import { View } from '../View';
import { Text } from '../../text/Text';
import { createFigma } from 'figma-api-stub';
const renderer = require('react-test-renderer');

describe('<View />', () => {
    beforeEach(() => {
        // @ts-ignore
        global.figma = createFigma({
            simulateErrors: true,
        });
    });

    it('View without children', () => {
        const tree = renderer.create(<View style={{ backgroundColor: '#ff0000' }} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('View with children', () => {
        const tree = renderer
            .create(
                <View style={{ flex: 1 }}>
                    <Text>text</Text>
                </View>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('View with children and rectangle styles', () => {
        const tree = renderer
            .create(
                <View style={{ flex: 1, backgroundColor: '#ff0000' }}>
                    <Text>text</Text>
                </View>
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('View with layoutAlign', () => {
        const tree = renderer
            .create(
                <View
                    style={{ width: 80, height: 40, backgroundColor: '#ffaa97' }}
                    layoutMode="HORIZONTAL"
                    layoutAlign="MAX"
                />
            )
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
});
