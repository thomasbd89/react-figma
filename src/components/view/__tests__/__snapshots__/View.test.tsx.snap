// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<View /> View with children 1`] = `
<frame
  backgrounds={
    Array [
      Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "r": 0,
        },
        "opacity": 0,
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      Object {
        "flex": 1,
      },
    ]
  }
>
  <text
    innerRef={
      Object {
        "current": undefined,
      }
    }
    loadedFont={null}
  >
    text
  </text>
</frame>
`;

exports[`<View /> View with children and rectangle styles 1`] = `
<frame
  backgrounds={
    Array [
      Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "r": 1,
        },
        "opacity": 1,
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Array [
      Object {
        "backgroundColor": "transparent",
      },
      Object {
        "backgroundColor": "#ff0000",
        "flex": 1,
      },
    ]
  }
>
  <text
    innerRef={
      Object {
        "current": undefined,
      }
    }
    loadedFont={null}
  >
    text
  </text>
</frame>
`;

exports[`<View /> View without children 1`] = `
<rectangle
  fills={
    Array [
      Object {
        "color": Object {
          "b": 0,
          "g": 0,
          "r": 1,
        },
        "opacity": 1,
        "type": "SOLID",
      },
    ]
  }
  innerRef={
    Object {
      "current": undefined,
    }
  }
  style={
    Object {
      "backgroundColor": "#ff0000",
    }
  }
/>
`;
